Adding MaxUnavailable to VMPool to ensure smoother update

aqilbeig opened the pull request:
### What this PR does
This PR would add MaxUnavailable as per [this](https://github.com/kubevirt/community/blob/main/design-proposals/vm-pool.md#virtualmachinepool-vmpool-api) design document which would ensure that all VMI doesn't get terminated simultaneously 

It would address https://github.com/kubevirt/kubevirt/issues/10328 and https://github.com/kubevirt/kubevirt/issues/14073

Current VMPool update strategy doesn't compare dataVolumeTempate changes [ part of VM template https://kubevirt.io/api-reference/main/definitions.html#_v1_virtualmachinespec ]; however, there could be scenarios where someone would like to update the source of the dataVolumetemplate to update the source image or e.g in case someone running the VMPool using custom image for Windows

datavolumetemplate changes doesn't reflect with VMI restart. it would only pick it with new VM creation and requires VM deletion rather than vmi deletion. VMPool would create the new vm with new datavolume source to ensure the number of replicas define in it

I am testing this changes further and would update the PR with test logs

Before this PR:

After this PR:
[logging-evidence.log](https://github.com/user-attachments/files/19318188/logging-evidence.log)

#### Release note
```release-note
Add maxUnavailable support to VirtualMachinePool

Adds support for controlling the maximum number of unavailable VMIs during pool updates. This helps maintain service availability by limiting how many VMs can be updated simultaneously.

- Defaults to 100% of total replicas for backward compatibility 
- Can be configured as percentage (e.g. "25%") or absolute number
- Prevents updates if current unavailable VMs exceed the limit
- Similar to Kubernetes Deployment's maxUnavailable behavior
```

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -889,9 +890,79 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+		// If we've reached the remaining updates limit, wait for some updates to complete
+		if processedCount >= remainingUpdates {
+			// Wait for at least one update to complete before continuing
+			select {
+			case err := <-errChan:
+				if err != nil {
+					return err
+				}
+				processedCount--
+			}
```
Is select necessary? Can't you just wait for errChan to return something?

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -889,9 +890,79 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+
 	for i := 0; i < len(vmUpdatedList); i++ {
+		// If we've reached the remaining updates limit, wait for some updates to complete
+		if processedCount >= remainingUpdates {
+			// Wait for at least one update to complete before continuing
+			select {
+			case err := <-errChan:
+				if err != nil {
+					return err
+				}
+				processedCount--
+			}
```
Shouldn't it break the loop in this case?

aqilbeig replied:
yes - it would .. 

if an update operation fails, there's likely an issue that would affect other updates as well. Continuing with more updates after one has failed could lead to an inconsistent state. The controller will requeue the work item, giving it another chance to process the updates in a future reconciliation.

0xFelix reviewed (commented):
Awesome work, some questions

aqilbeig commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter.go:
```diff
@@ -131,6 +134,38 @@ func ValidateVMPoolSpec(ar *admissionv1.AdmissionReview, field *k8sfield.Path, p
+			if !strings.HasSuffix(spec.MaxUnavailable.StrVal, "%") {
+				causes = append(causes, metav1.StatusCause{
+					Type:    metav1.CauseTypeFieldValueInvalid,
+					Message: fmt.Sprintf("maxUnavailable percentage must end with %%"),
```
@0xFelix There's no need for fmt.Sprintf in this case since there are no format specifiers or variables being interpolated into the string. I used this just to maintain the consistency as it has been used with the static string in other places as well like https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter.go#L112...

let me update it.. thanks

aqilbeig commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter.go:
```diff
@@ -131,6 +134,38 @@ func ValidateVMPoolSpec(ar *admissionv1.AdmissionReview, field *k8sfield.Path, p
 
+	// Validate MaxUnavailable if set
+	if spec.MaxUnavailable != nil {
+		if spec.MaxUnavailable.Type == intstr.String {
```
https://pkg.go.dev/k8s.io/apimachinery/pkg/util/intstr#IntOrString

there can only be two types for intstr.Type:

- intstr.String - for percentage values (e.g. "25%")
- intstr.Int - for absolute numbers

This is enforced by the Kubernetes API machinery itself, as IntOrString is a built-in Kubernetes type from k8s.io/apimachinery/pkg/util/intstr

We can still add an default case in case needed

0xFelix replied:
If a user puts in some value other than what the `intstr.String` or `intstr.Int` consts contain, is it still possible to unmarshal that into the object we're checking here?

aqilbeig replied:
@0xFelix 
While apply vmpool spec with maxUnavailable as -> false [boolean]
```
The VirtualMachinePool "vm-pool-cirros" is invalid:
* spec.maxUnavailable: Invalid value: "boolean": spec.maxUnavailable in body must be of type integer,string: "boolean"
* <nil>: Invalid value: "": "spec.maxUnavailable" must validate at least one schema (anyOf)
* spec.maxUnavailable: Invalid value: "boolean": spec.maxUnavailable in body must be of type integer: "boolean"
```

While apply vmpool spec with maxUnavailable as -> 0.05
```
❯ k apply -f vmpool-ciros.yaml
The VirtualMachinePool "vm-pool-cirros" is invalid:
* spec.maxUnavailable: Invalid value: "number": spec.maxUnavailable in body must be of type integer,string: "number"
* <nil>: Invalid value: "": "spec.maxUnavailable" must validate at least one schema (anyOf)
* spec.maxUnavailable: Invalid value: "number": spec.maxUnavailable in body must be of type integer: "number"
```

0xFelix commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter.go:
```diff
@@ -126,11 +129,43 @@ func ValidateVMPoolSpec(ar *admissionv1.AdmissionReview, field *k8sfield.Path, p
+				causes = append(causes, metav1.StatusCause{
+					Type:    metav1.CauseTypeFieldValueInvalid,
+					Message: "maxUnavailable percentage must end with %",
+					Field:   field.Child("maxUnavailable").String(),
```
You could dedup the field, define it only once.

0xFelix commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter.go:
```diff
@@ -126,11 +129,43 @@ func ValidateVMPoolSpec(ar *admissionv1.AdmissionReview, field *k8sfield.Path, p
 		})
 	}
 
+	// Validate MaxUnavailable if set
```
IMO the code speaks for itself here, we don't need the extra comments.

0xFelix commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter_test.go:
```diff
@@ -136,6 +137,70 @@ var _ = Describe("Validating Pool Admitter", func() {
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{"match": "me"},
+					},
+					Spec: v1.VirtualMachineSpec{
```
Can you simplify the VM spec in this case?

0xFelix replied:
Can you reduce the VirtualMachinePoolSpec to the minimum that is required to trigger the expected rejection?

aqilbeig replied:
these are the required fields - i tried removing selector and VirtualMachineTemplate - however tests are failing with :
```
• [FAILED] [0.000 seconds]
Validating Pool Admitter reject invalid VirtualMachineInstance spec [It] with invalid maxUnavailable percentage
pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter_test.go:140

  Timeline >>
  {"component":"","level":"info","msg":"rejected vmi admission","pos":"webhooks.go:55","timestamp":"2025-03-11T17:02:45.480742Z"}
  [FAILED] in [It] - pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter_test.go:91 @ 03/11/25 17:02:45.48
  << Timeline

  [FAILED] Expected
      <string>: spec.template
  to equal
      <string>: spec.maxUnavailable
  In [It] at: pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter_test.go:91 @ 03/11/25 17:02:45.48
------------------------------
• [FAILED] [0.000 seconds]
Validating Pool Admitter reject invalid VirtualMachineInstance spec [It] with invalid maxUnavailable integer
pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter_test.go:150

  Timeline >>
  {"component":"","level":"info","msg":"rejected vmi admission","pos":"webhooks.go:55","timestamp":"2025-03-11T17:02:45.481125Z"}
  [FAILED] in [It] - pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter_test.go:89 @ 03/11/25 17:02:45.481
  << Timeline

  [FAILED] Expected
      <[]v1.StatusCause | len:3, cap:4>: [
          {
              Type: "FieldValueInvalid",
              Message: "RunStrategy must be specified",
              Field: "spec.virtualMachineTemplate.spec.running",
          },
          {
              Type: "FieldValueInvalid",
              Message: "selector does not match labels.",
              Field: "spec.selector",
          },
          {
              Type: "FieldValueInvalid",
              Message: "maxUnavailable must be a positive number",
              Field: "spec.maxUnavailable",
          },
      ]
  to have length 1
  In [It] at: pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter_test.go:89 @ 03/11/25 17:02:45.481
```

0xFelix commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter_test.go:
```diff
@@ -136,6 +137,70 @@ var _ = Describe("Validating Pool Admitter", func() {
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{"match": "me"},
+					},
+					Spec: v1.VirtualMachineSpec{
```
Same

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
Same about the comments in this file, most of the code speaks for itself (or it should be)

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -889,9 +890,76 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+	for _, vm := range vmUpdatedList {
+		vmiKey := controller.NamespacedKey(vm.Namespace, vm.Name)
+		obj, exists, _ := c.vmiStore.GetByKey(vmiKey)
+		if !exists {
```
Isn't that going to be an issue? In that case we should try to get the VM with the client and eventually return with an error if it cannot be found?

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -889,9 +890,76 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+
 	for i := 0; i < len(vmUpdatedList); i++ {
+		// If we've reached the remaining updates limit, wait for some updates to complete
+		if processedCount >= remainingUpdates {
```
Can you explain what we need this for? IIUC the existing code uses routines to parallelize this and to speed it up? Why should it wait now?

0xFelix replied:
Got it. Can you add a functional test for that please?

0xFelix reviewed (commented):
Please update the PR description to include a release note.

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,6 +833,33 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
 	return nil, false
 }
 
+func (c *Controller) getUnavailableVMICount(pool *poolv1.VirtualMachinePool, vms []*virtv1.VirtualMachine) (int, error) {
```
`pool` is unused?

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -889,9 +917,55 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
 	errChan := make(chan error, len(vmUpdatedList))
+	unavailableCount, err := c.getUnavailableVMICount(pool, vmUpdatedList)
+	if err != nil {
+		errChan <- err
```
That looks weird, why bother writing to `errChan`?

aqilbeig replied:
yeah. this is wrong .. I put this for testing purpose and forgot to remove it - idea was not to trigger the update of vmpool which already has the unhealthy vms or vmis 

Use case:
- Someone create the vmpool with wrong configuration - and no vmi running for it. we should avoid triggering update in that case

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -889,9 +917,55 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+		unavailableCount, err := c.getUnavailableVMICount(pool, vmUpdatedList)
+		if err != nil {
+			errChan <- err
+			return err
```
This for loop does not run in a separate routine. Why bother writing to the channel?

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -889,9 +917,55 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+	}
+
 	for i := 0; i < len(vmUpdatedList); i++ {
+		unavailableCount, err := c.getUnavailableVMICount(pool, vmUpdatedList)
```
I suspect that if we add a functional test, this will produce many flakes. Can you please implement a functional test to prove it works / it doesn't work?

aqilbeig replied:
yes.. working on that

aqilbeig commented on pkg/virt-controller/watch/pool/pool_test.go:
```diff
@@ -817,3 +803,27 @@ func DefaultPool(replicas int32) (*poolv1.VirtualMachinePool, *v1.VirtualMachine
 }
+
+func createReadyVMI(vm *v1.VirtualMachine, poolRevision *appsv1.ControllerRevision) *v1.VirtualMachineInstance {
+	vmi := api.NewMinimalVMI(vm.Name)
```
@0xFelix I am just moving the lines https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool_test.go#L340-L354 to a function for reusability 

I could update it to:
```
vmi := libvmi.New(
    libvmi.WithName(vm.Name),
    libvmi.WithSpec(vm.Spec.Template.Spec),
)
```
would that work?

aqilbeig commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -888,88 +960,128 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
-			defer wg.Done()
-			vm := vmUpdatedList[idx]
+	// Handle unhealthy VMIs first to rollover any changes to the VMI spec in case last update failed
+	if err := c.handleUnhealthyVMIs(pool, vmUpdatedList); err != nil {
```
This function is handling the update for use cases - where existing VMPool is unhealthy with multiple VMs/VMIs are not ready as the launcher pod is crashing with some reason

How to test this functionality:
- Create a vmpool with below specs:
e.g.
```
apiVersion: pool.kubevirt.io/v1alpha1
kind: VirtualMachinePool
metadata:
  name: vm-pool-cirros
  namespace: ns-beig
spec:
  replicas: 10
  selector:
    matchLabels:
      kubevirt.io/vmpool: vm-pool-cirros
  virtualMachineTemplate:
    metadata:
      creationTimestamp: null
      labels:
        kubevirt.io/vmpool: vm-pool-cirros
    spec:
      runStrategy: Always
      template:
        metadata:
          creationTimestamp: null
          labels:
            kubevirt.io/vmpool: vm-pool-cirros
        spec:
          domain:
            devices:
              disks:
              - disk:
                  bus: virtio
                name: containerdisk
            resources:
              requests:
                memory: 128Mi
          terminationGracePeriodSeconds: 0
          volumes:
          - containerDisk:
              image: kubevirt/cirros-container-disk-demo:v0.36.3
            name: containerdisk 
```

- Update the VMPool with non existent container Images i.e. change the image to kubevirt/cirros-container-disk-demo:v0.36.10
- This would trigger the update - however, as per the maxUnavailable it will restart two VMIs and new VMIs would error out with Init:ImagePullBackOff
- At this time, only two VMI would be fail with this and the update won't update further VMIs to respect maxUnavailableInt
- Update the VMpool again with right image i.e. 0.36.4

this function would trigger the update for VMIs in notREady state and let other VMis to update with below function

0xFelix replied:
Can you break down the logic of `proactiveUpdate` into smaller methods?

aqilbeig commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -888,88 +960,128 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+		for {
+			select {
+			case <-timeout:
+				return fmt.Errorf("timeout as the new VMIs are not coming online after 1 minutes")
```
This is to handle the use case - where someone a bad update to existing VMPool. e.g. lets say we have a healthy vmpool with below specs:
```
apiVersion: pool.kubevirt.io/v1alpha1
kind: VirtualMachinePool
metadata:
  name: vm-pool-cirros
  namespace: ns-beig
spec:
  replicas: 10
  selector:
    matchLabels:
      kubevirt.io/vmpool: vm-pool-cirros
  virtualMachineTemplate:
    metadata:
      creationTimestamp: null
      labels:
        kubevirt.io/vmpool: vm-pool-cirros
    spec:
      runStrategy: Always
      template:
        metadata:
          creationTimestamp: null
          labels:
            kubevirt.io/vmpool: vm-pool-cirros
        spec:
          domain:
            devices:
              disks:
              - disk:
                  bus: virtio
                name: containerdisk
            resources:
              requests:
                memory: 128Mi
          terminationGracePeriodSeconds: 0
          volumes:
          - containerDisk:
              image: kubevirt/cirros-container-disk-demo:v0.36.3
            name: containerdisk 
```
- Update the vmpool with non-existent container image like 0.36.12
- Update will stuck as soon as initial two VMIs go in imagePullBackOff as we cannot proceed further due to maxUnavailable 
- this code will ensure timeout after checking for a minute for new VMIs to come healthy

aqilbeig commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -888,88 +960,128 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+			case <-timeout:
+				return fmt.Errorf("timeout as the new VMIs are not coming online after 1 minutes")
+			default:
+				time.Sleep(1 * time.Second)
```
There is delay in the actual VMI deletion/restart happening after the code is triggering deleteOptions during proactiveUpdateTypeRestart. Due to which we are getting unavailableCount as zero for the first few iterations

aqilbeig commented:
I have test all the use-cases - here is the logs for one of the use case
[Uploading logging-evidence.log…]()

aqilbeig commented:
@0xFelix request you to please review the pr?

0xFelix commented on staging/src/kubevirt.io/api/pool/v1alpha1/types.go:
```diff
@@ -112,6 +113,10 @@ type VirtualMachinePoolSpec struct {
+
+	// (Defaults to 25%) Integer or string pointer, that when set represents either a percentage or number of VMs in a pool that can be unavailable (ready condition false) at a time during automated update.
+	// +optional
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,3,opt,name=maxUnavailable"`
```
What's the current behavior without this setting? Can you default it to a value that keeps the current behavior?

aqilbeig replied:
@0xFelix Current behaviour is to restart all VMIs without any maxUnavailable? that won't be a nicer approach as that would cause downtime to any vmpool. Also, current behaviour could be achievable by setting maxUnavailable=100%. Current behaviour is also not as per the initial design proposal of VMPool which suggested to have maxUnavailable defaults to 25%

Do let me know in case you still feel that we should go with 100% as the default.. I'll update the code.

0xFelix replied:
Please go with 100% as the default. Although the proposal states otherwise it was working for an already long time like this.

aqilbeig replied:
ok.. i'll update it..

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,6 +838,71 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
+}
+
+func (c *Controller) handleUnhealthyVMIs(pool *poolv1.VirtualMachinePool, vms []*virtv1.VirtualMachine) error {
+	notReadyVMs := c.filterNotReadyVMs(vms)
```
Might be able to use `slices.DeleteFunc` here instead of the custome `filterVMs`

aqilbeig replied:
@0xFelix slices.DeleteFunc would delete the object directly from the original slice which would delete all the VM object from the original slice in this case. I have to create a copy of vms object to pass to slices.deleteFunc which would add more line of code than this function itself

0xFelix replied:
ACK

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,6 +838,71 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
+
+func (c *Controller) handleUnhealthyVMIs(pool *poolv1.VirtualMachinePool, vms []*virtv1.VirtualMachine) error {
+	notReadyVMs := c.filterNotReadyVMs(vms)
+	if len(notReadyVMs) == 0 {
+		return nil
+	}
```
Can be dropped, loop should do nothing if it's empty.

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,6 +838,71 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
+					return err
+				}
+			} else if updateType == proactiveUpdateTypeVMDelete {
+				foreGround := metav1.DeletePropagationForeground
```
use `pointer.P` for value of `PropagationPolicy`

0xFelix replied:
You can inline `foreGround`

0xFelix reviewed (commented):
Can you add a functional test please? See `/tests`.

aqilbeig commented:
@0xFelix could you please review it again and check the comments?

0xFelix commented:
@aqilbeig Will do a more thorough review next week.

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,6 +838,68 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
+			}
+			obj = vmi
+		}
+		vmi := obj.(*virtv1.VirtualMachineInstance)
```
Might be good to check that the type assertion was successful. (check `ok` is true)

aqilbeig replied:
@0xFelix 
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L164
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L164
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L1126

same has been used at multiple places in pool.go.. and there are multiple references for the same https://github.com/search?q=repo%3Akubevirt%2Fkubevirt%20%22vmi%20%3A%3D%20obj.(*virtv1.VirtualMachineInstance)%22&type=code in other files too. we are not checking the assertion at anywhere. I don't think this would be needed here

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,6 +838,68 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
+			if err != nil {
+				return err
+			}
+			if updateType == proactiveUpdateTypeRestart {
```
You could construct a `metav1.DeleteOptions{}` in the conditionals and then have a single block that does the delete, effectively de-duping the code.

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -887,89 +955,138 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+	}
+
+	for i := range vmUpdatedList {
+		timeout := time.After(2 * time.Minute)
```
Can the controller return and requeue instead of blocking?

aqilbeig replied:
this is required for cases where new VMI is taking time to come up healthy like in case of data volume changes. also, it would also ensure to timeout for cases where new VMIs are not coming up healthy due to any reason so that it could timeout with an error and requeue it instead of waiting it indefinitely

aqilbeig replied:
I have shared the details earlier -> https://github.com/kubevirt/kubevirt/pull/14183#discussion_r2000884786

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -887,89 +955,138 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+	return maxUnavailableInt, nil
+}
+
+func (c *Controller) proactiveUpdate(pool *poolv1.VirtualMachinePool, vmUpdatedList []*virtv1.VirtualMachine) error {
```
There's way too much happening in this single function. Can you please break it down into smaller functions?

0xFelix commented on pkg/virt-controller/watch/pool/pool_test.go:
```diff
@@ -771,6 +743,102 @@ var _ = Describe("Pool", func() {
 			Entry("append index if set to true", pointer.P(true)),
 		)
+
+		It("should respect maxUnavailable limit during proactive updates", func() {
```
Can you test the backoff part of maxUnavailable VMs in unit tests?

0xFelix commented on tests/pool_test.go:
```diff
@@ -610,6 +611,72 @@ var _ = Describe("[sig-compute]VirtualMachinePool", decorators.SigCompute, func(
+
+		By("Verifying maxUnavailable strategy")
+		Eventually(func() error {
+			vmis, err := virtClient.VirtualMachineInstance(newPool.Namespace).List(context.Background(), metav1.ListOptions{
+				LabelSelector: labelSelectorToString(newPool.Spec.Selector),
+			})
+			if err != nil {
+				return err
+			}
+
+			updatedCount := 0
+			for _, vmi := range vmis.Items {
+				if _, ok := vmi.Labels[newLabelKey]; ok {
+					updatedCount++
+				}
+			}
+
+			deletingCount := 0
+			for _, vmi := range vmis.Items {
+				if vmi.DeletionTimestamp != nil {
+					deletingCount++
+				}
+			}
+
+			if deletingCount+updatedCount > 1 {
+				return fmt.Errorf("Too many VMIs being updated simultaneously. Deleting: %d, Updated: %d", deletingCount, updatedCount)
+			}
+
+			if updatedCount < 3 {
+				return fmt.Errorf("Not all VMIs updated yet. Updated: %d/3", updatedCount)
+			}
+
+			return nil
```
Can you change this to expects?

aqilbeig replied:
No, we cannot use Expect instead of Eventually in this case. Here's why:

- Eventually is specifically designed for asynchronous testing scenarios where we need to wait for a condition to be met over time. It repeatedly checks the condition until it succeeds or times out.
- Expect is for synchronous assertions where we want to check a condition immediately. It doesn't retry or wait.

In this specific test case, we're testing the maxUnavailable strategy which involves:

- Waiting for VMIs to be updated one at a time
- Checking that no more than one VMI is being updated simultaneously
- Verifying that all VMIs eventually get updated

0xFelix replied:
Sounds like some AI hallucination.

I'm asking you to replace returning errors by using `Expect`s instead.

```
Eventually(func(g Gomega) {
	g.Expect(true).To(BeTrue())
	g.Expect("something).To(Equal("something"))
}, 180*time.Second, 1*time.Second).Should(Succeed())
```

aqilbeig replied:
yeah i used chatGPT to identity what would be best here.. and got that response.. let me update it in case needed

aqilbeig commented:
thanks a lot @0xFelix for reviewing this PR multiple times..

aqilbeig commented:
@0xFelix could you please review it again? Is there a way I could sync with you on slack or teams just to close any remaining concerns to move forward with this PR?

0xFelix commented:
@aqilbeig You can contact me on the Kubernetes slack.

aqilbeig commented:
@0xFelix could you please review the pr?

aqilbeig commented:
Logs with maxUnavailable defined:
[vmpool-update.txt](https://github.com/user-attachments/files/19669550/vmpool-update.txt)


These are the logs with maxUnavailable as default i.e. when someone doesn't mention or use maxUnavailable in there specs - it is terminating all at once like you suggested
[vmpool-update-default.txt](https://github.com/user-attachments/files/19669551/vmpool-update-default.txt)


Vmpool Spec used to test the same:
```
apiVersion: pool.kubevirt.io/v1alpha1
kind: VirtualMachinePool
metadata:
  name: vm-pool-cirros
  namespace: ns-beig
spec:
  replicas: 10
  selector:
    matchLabels:
      kubevirt.io/vmpool: vm-pool-cirros
  virtualMachineTemplate:
    metadata:
      creationTimestamp: null
      labels:
        kubevirt.io/vmpool: vm-pool-cirros
    spec:
      runStrategy: Always
      template:
        metadata:
          creationTimestamp: null
          labels:
            kubevirt.io/vmpool: vm-pool-cirros
        spec:
          domain:
            devices:
              disks:
              - disk:
                  bus: virtio
                name: containerdisk
            resources:
              requests:
                memory: 128Mi
          terminationGracePeriodSeconds: 0
          volumes:
          - containerDisk:
              image: kubevirt/cirros-container-disk-demo:v0.36.4
            name: containerdisk 
``

aqilbeig commented:
@0xFelix Could you please review and approve this PR?

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -545,10 +552,10 @@ func (c *Controller) scaleIn(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virt
 
 			foreGround := metav1.DeletePropagationForeground
-			err := c.clientset.VirtualMachine(vm.Namespace).Delete(context.Background(), vm.Name, metav1.DeleteOptions{PropagationPolicy: &foreGround})
+			err := c.clientset.VirtualMachine(vm.Namespace).Delete(context.Background(), vm.Name, metav1.DeleteOptions{PropagationPolicy: pointer.P(foreGround)})
```
```suggestion
			err := c.clientset.VirtualMachine(vm.Namespace).Delete(context.Background(), vm.Name, metav1.DeleteOptions{PropagationPolicy: pointer.P(metav1.DeletePropagationForeground)})
```

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -545,10 +552,10 @@ func (c *Controller) scaleIn(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virt
 			if err != nil {
 				c.expectations.DeletionObserved(poolKey, controller.VirtualMachineKey(vm))
-				c.recorder.Eventf(pool, k8score.EventTypeWarning, common.FailedDeleteVirtualMachineReason, "Error deleting virtual machine %s: %v", vm.ObjectMeta.Name, err)
+				c.recorder.Eventf(pool, k8score.EventTypeWarning, common.FailedDeleteVirtualMachineReason, "Error deleting virtual machine %s/%s: %v", vm.ObjectMeta.Name, vm.Namespace, err)
```
```suggestion
				c.recorder.Eventf(pool, k8score.EventTypeWarning, common.FailedDeleteVirtualMachineReason, "Error deleting virtual machine %s/%s: %v", vm.Namespace, vm.Name, err)
```

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,6 +839,55 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
+	if vmi.DeletionTimestamp != nil || vmi.Status.Phase != virtv1.Running {
+		return false
+	}
+	readyCondition := controller.NewVirtualMachineInstanceConditionManager().GetCondition(vmi, virtv1.VirtualMachineInstanceReady)
```
Use `HasConditionWithStatus` instead.

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,6 +839,55 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
+	if vmi.DeletionTimestamp != nil || vmi.Status.Phase != virtv1.Running {
+		return false
+	}
+	readyCondition := controller.NewVirtualMachineInstanceConditionManager().GetCondition(vmi, virtv1.VirtualMachineInstanceReady)
+	return readyCondition != nil && readyCondition.Status == k8score.ConditionTrue
```
```suggestion
	return controller.NewVirtualMachineInstanceConditionManager().HasConditionWithStatus(vmi, virtv1.VirtualMachineInstanceReady, k8score.ConditionTrue)
```

0xFelix replied:
What you are doing in this helper is redundant. You can replace this helper with `controller.NewVirtualMachineInstanceConditionManager().HasConditionWithStatus(vmi, virtv1.VirtualMachineInstanceReady, k8score.ConditionTrue)` entirely.

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,6 +839,55 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
+	notReadyVMs := c.filterNotReadyVMs(vms)
+
+	for _, vm := range notReadyVMs {
+		vmiKey := controller.NamespacedKey(vm.Namespace, vm.Name)
+		obj, exists, _ := c.vmiStore.GetByKey(vmiKey)
```
```suggestion
		obj, exists, _ := c.vmiStore.GetByKey(controller.NamespacedKey(vm.Namespace, vm.Name))
```

0xFelix replied:
What about `err`?

aqilbeig replied:
took the reference from https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L900 and https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L1124

0xFelix replied:
I see but it still looks suspicious to me, we should check the `err`.

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -887,89 +943,82 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
-			if err != nil {
-				errChan <- err
-				return
+	var maxUnavailableInt int
```
For consistency

```suggestion
	maxUnavailableInt := 0
```

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -887,89 +943,82 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+
+		vm := vmUpdatedList[i]
+		vmiKey := controller.NamespacedKey(vm.Namespace, vm.Name)
+		obj, exists, _ := c.vmiStore.GetByKey(vmiKey)
```
What about `err`?

aqilbeig replied:
took the reference from https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L1124 and https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L900

0xFelix replied:
Same answer as above

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -887,89 +943,82 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+		vm := vmUpdatedList[i]
+		vmiKey := controller.NamespacedKey(vm.Namespace, vm.Name)
+		obj, exists, _ := c.vmiStore.GetByKey(vmiKey)
+		if !exists {
+			return nil
+		}
```
Shouldn't this be a continue?

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -1048,6 +1101,13 @@ func (c *Controller) isOutdatedVMI(vm *virtv1.VirtualMachine, vmi *virtv1.Virtua
+	currentDataVolumeTemplates := poolSpecRevisionForVMI.VirtualMachineTemplate.Spec.DataVolumeTemplates
+
+	// If DataVolumeTemplates differ, we need to delete the VM to ensure proper data volume handling
+	if !equality.Semantic.DeepEqual(currentDataVolumeTemplates, expectedDataVolumeTemplates) {
```
Isn't that an additional change on top of the max unavailable feature?

aqilbeig replied:
not directly related to maxUnavailable - however this was missing as this was not tacking the changes in DV

0xFelix replied:
Can we keep it separate, at least in a separate commit, better in a separate PR?

aqilbeig replied:
ok.. i'll add these in a separate commit

aqilbeig replied:
Added as a separate commit https://github.com/kubevirt/kubevirt/pull/14183/commits/357c6f2f71ffeb4a9e00f153d9334fb349c55d0d

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -1353,3 +1413,44 @@ func (c *Controller) execute(key string) error {
+		err = c.clientset.VirtualMachineInstance(vm.ObjectMeta.Namespace).Delete(context.Background(), vmi.ObjectMeta.Name, metav1.DeleteOptions{})
+		log.Log.Object(pool).Infof("Proactive update of VM %s/%s by deleting outdated VMI", vm.Namespace, vm.Name)
+	case proactiveUpdateTypeVMDelete:
+		foreGround := metav1.DeletePropagationForeground
```
Same as above

0xFelix replied:
Why use foreground deletion in the first place?

aqilbeig replied:
i took the reference from https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L547 .. I'll modify it though

0xFelix replied:
We could try without. Can you search the commit history / PRs adding this for an explanation?

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -1353,3 +1413,44 @@ func (c *Controller) execute(key string) error {
+
+	switch updateType {
+	case proactiveUpdateTypeRestart:
+		err = c.clientset.VirtualMachineInstance(vm.ObjectMeta.Namespace).Delete(context.Background(), vmi.ObjectMeta.Name, metav1.DeleteOptions{})
```
VMs have a `Restart` subresource, use that?

aqilbeig replied:
Current implementation is using delete only https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L919 - could you please share the benefits of using restart over delete? is the current implementation inaccurate?

0xFelix replied:
IIRC it should do that, can you try?

0xFelix commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -1353,3 +1413,44 @@ func (c *Controller) execute(key string) error {
+		err = c.clientset.VirtualMachine(vm.ObjectMeta.Namespace).Delete(context.Background(), vm.ObjectMeta.Name, metav1.DeleteOptions{PropagationPolicy: pointer.P(foreGround)})
+		log.Log.Object(pool).Infof("Proactive update of VM %s/%s by deleting VM due to data volume changes", vm.Namespace, vm.Name)
+	case proactiveUpdateTypePatchRevisionLabel:
+		patch := []map[string]interface{}{
```
Use `patchset` to construct the patch please.

0xFelix commented on pkg/virt-controller/watch/pool/pool_test.go:
```diff
@@ -148,6 +150,18 @@ var _ = Describe("Pool", func() {
 			controller.queue = mockQueue
 
+			stop := make(chan struct{})
+			go vmiInformer.Run(stop)
```
@fossedihelm Please remind me, is that still the way to go?

0xFelix replied:
Ping

aqilbeig replied:
ok i'll update.. there are many tests using the same - i took the reference from there like - https://github.com/search?q=repo%3Akubevirt%2Fkubevirt%20go%20vmiInformer.Run(stop)&type=code

aqilbeig replied:
I think I could just remove this as these tests are already using the fake client

0xFelix commented on pkg/virt-controller/watch/pool/pool_test.go:
```diff
@@ -800,9 +958,12 @@ func PoolFromVM(name string, vm *v1.VirtualMachine, replicas int32) *poolv1.Virt
-	vm := watchtesting.VirtualMachineFromVMI(vmi.Name, vmi, true)
+	vmi := libvmi.New(
+		libvmi.WithNamespace(k8sv1.NamespaceDefault),
+		libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+		libvmi.WithNetwork(v1.DefaultPodNetwork()),
```
Do these tests really require this network?

0xFelix commented on tests/pool_test.go:
```diff
@@ -610,6 +611,66 @@ var _ = Describe("[sig-compute]VirtualMachinePool", decorators.SigCompute, func(
+
+	It("should respect maxUnavailable strategy during updates", func() {
+		newPool := newVirtualMachinePool()
+		doScale(newPool.ObjectMeta.Name, 3)
```
```suggestion
		doScale(newPool.Name, 3)
```

0xFelix replied:
and also below

aqilbeig replied:
@0xFelix this has been used like this only at multiple places in this test like https://github.com/kubevirt/kubevirt/blob/main/tests/pool_test.go#L206
https://github.com/kubevirt/kubevirt/blob/main/tests/pool_test.go#L223

and that is the reason i am using the same for consistency. Do we want to change the whole file references?

0xFelix replied:
Keep changes to this test for now, but I'd still change it.

0xFelix commented on tests/pool_test.go:
```diff
@@ -610,6 +611,66 @@ var _ = Describe("[sig-compute]VirtualMachinePool", decorators.SigCompute, func(
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Verifying maxUnavailable strategy")
+		Eventually(func() {
```
Shouldn't this be more like a Consistently? Or can you manually track the calls to update VMs? This looks racy to me.

aqilbeig replied:
yes .. good point.. this could hit a race condition.. Consistently seems more relevant here as it will repeatedly check it. I'll change the eventually to consistently -

```
		Consistently(func() {
			vmis, err := virtClient.VirtualMachineInstance(newPool.Namespace).List(context.Background(), metav1.ListOptions{
				LabelSelector: labelSelectorToString(newPool.Spec.Selector),
			})
			Expect(err).ToNot(HaveOccurred())

			unavailableCount := 0
			for _, vmi := range vmis.Items {
				if vmi.DeletionTimestamp != nil || vmi.Status.Phase != v1.Running {
					unavailableCount++
				}
			}

			Expect(unavailableCount).To(BeNumerically("<=", 1),
				fmt.Sprintf("Too many unavailable VMIs: %d (max allowed: 1)", unavailableCount))

			updatedCount := 0
			for _, vmi := range vmis.Items {
				if _, ok := vmi.Labels[newLabelKey]; ok {
					updatedCount++
				}
			}

			Expect(updatedCount).To(Equal(3),
				fmt.Sprintf("Not all VMIs updated yet. Updated: %d/3", updatedCount))
		}, 5*time.Second, 1*time.Second).Should(Succeed())
```

would this be fine?

0xFelix replied:
Consistently might be better, but can you instead manually track the update calls instead? That sounds like the safest way to me.

0xFelix reviewed (commented):
There are still some issues to fix.

aqilbeig commented:
@0xFelix could you please check my responses.. I'll update the PR post to avoid any further commits

aqilbeig commented:
@0xFelix I have updated the PR and addressed almost of your concern. Could you please review it and move forward this?

0xFelix commented on pkg/virt-controller/watch/pool/pool_test.go:
```diff
@@ -156,7 +170,25 @@ var _ = Describe("Pool", func() {
 			virtClient.EXPECT().VirtualMachinePool(testNamespace).Return(fakeVirtClient.PoolV1alpha1().VirtualMachinePools(testNamespace)).AnyTimes()
 
+			// Add VMI Get reactor to handle VMI Get requests
+			fakeVirtClient.Fake.PrependReactor("get", "virtualmachineinstances", func(action k8stesting.Action) (handled bool, obj runtime.Object, err error) {
```
Why should the client get the vmi from the store? It could not exist in the informers store but it could exist on the API server.

aqilbeig replied:
@0xFelix I didn't understand. this is a mock test. Would there be an api-server with mock test?

aqilbeig replied:
@0xFelix @fossedihelm https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool_test.go#L113 these tests use addVMi which add the VMI to the vmistore - in case - I change it to use fake client - don't we have to add vmi to fake virt client store too...

I could change it to :
```
			fakeVirtClient.Fake.PrependReactor("get", "virtualmachineinstances", func(action k8stesting.Action) (handled bool, obj runtime.Object, err error) {
				getAction, ok := action.(k8stesting.GetAction)
				if !ok {
					return false, nil, nil
				}
				vmi, err := fakeVirtClient.KubevirtV1().VirtualMachineInstances(getAction.GetNamespace()).Get(context.Background(), getAction.GetName(), metav1.GetOptions{})
				if err != nil {
					return true, nil, err
				}
				return true, vmi, nil
			})
```

However, fakeVirtclient doesn't have that as current functionality adding it to vmi store.. Could you please help me out here like how should i modify this to use the fake client

0xFelix commented on pkg/virt-controller/watch/pool/pool_test.go:
```diff
@@ -771,6 +774,161 @@ var _ = Describe("Pool", func() {
+
+			addPool(pool)
+			// Create 4 VMs with their VMIs
+			for i := 0; i < 4; i++ {
```
```suggestion
			for i := range 4 {
```

0xFelix commented on tests/pool_test.go:
```diff
@@ -610,6 +611,70 @@ var _ = Describe("[sig-compute]VirtualMachinePool", decorators.SigCompute, func(
+				}
+
+				if updatedCount == replicas {
+					close(done)
```
If done then the Consistently should bail too?

0xFelix replied:
Closing the chan multiple times will panic?

0xFelix commented on tests/pool_test.go:
```diff
@@ -610,6 +611,70 @@ var _ = Describe("[sig-compute]VirtualMachinePool", decorators.SigCompute, func(
+					close(done)
+				}
+				return nil
+			}, 120*time.Second, 1*time.Second).Should(Succeed(), "maxUnavailable constraint was violated during the update process")
+		}()
+
+		By("Waiting for the update process to complete")
+		Eventually(done, 150*time.Second).Should(BeClosed(), "update process did not complete within the timeout")
```
Timeouts do not match?

aqilbeig replied:
yeah i wanted to keep more time than consistently block to give updates a little buffer time to finish

aqilbeig commented:
> Can you please squash your commits where it makes sense?
> 
> Also the logic in the func test still looks a bit odd to me. Why is there a need for a channel? After Consistently is done you could assert immediately if the operation is done. That would also save you from adding the additional routine.

i was thinking to run it in routine so that i could track whether all updates has been completed or not as i was thinking of the scenario where upgrade is not finished after the consistently block. however, that shouldn't happen as it is creating only 4 replicas so that shouldn't be a problem. let me simplify it then..

aqilbeig commented:
@0xFelix i have simplified the functional test and tested that too in my local env. Also, i have squashed the commits too. Request you to please move this PR forward now

0xFelix commented on tests/pool_test.go:
```diff
@@ -610,6 +611,68 @@ var _ = Describe("[sig-compute]VirtualMachinePool", decorators.SigCompute, func(
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Verifying maxUnavailable constraint throughout the update process")
+		Consistently(func() error {
```
I'm a little concerned this test will flake if the test cluster is too slow. Let's give it a go to see how good it works.

0xFelix reviewed (commented):
/test all

aqilbeig commented:
@fossedihelm could you please review this PR?

fossedihelm commented:
@aqilbeig @0xFelix looking :eyes:

fossedihelm commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -887,89 +944,84 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
+		if err != nil {
+			return 0, fmt.Errorf("invalid maxUnavailable percentage: %v", err)
+		}
+		maxUnavailableInt = int(float64(totalReplicas) * float64(percentage) / 100.0)
```
pecentage cannot be float:
```suggestion
		maxUnavailableInt = int(totalReplicas) * int(percentage) / 100
```

fossedihelm reviewed (commented):
Thank you!
Some question below. I admit that I have not finished the entire review, but I prefer going step by step

aqilbeig commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -1353,3 +1405,52 @@ func (c *Controller) execute(key string) error {
+		log.Log.Object(pool).Infof("Proactive update of VM %s/%s by deleting outdated VMI", vm.Namespace, vm.Name)
+	case proactiveUpdateTypePatchRevisionLabel:
+		patchSet := patch.New()
+		vmiCopy := vmi.DeepCopy()
```
@fossedihelm This is the existing code i am using -> https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L927-L937

aqilbeig commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -1353,3 +1405,52 @@ func (c *Controller) execute(key string) error {
+		vmiCopy.Labels[virtv1.VirtualMachinePoolRevisionName] = revisionName
+
+		if vmi.Labels == nil {
+			patchSet.AddOption(patch.WithAdd("/metadata/labels", vmi.Labels))
```
@fossedihelm Using the existing code only https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L941 ..

aqilbeig commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -1353,3 +1405,52 @@ func (c *Controller) execute(key string) error {
+			patchSet.AddOption(patch.WithAdd("/metadata/labels", vmi.Labels))
+		} else {
+			patchSet.AddOption(
+				patch.WithTest("/metadata/labels", vmiCopy.Labels),
+				patch.WithReplace("/metadata/labels", vmi.Labels),
```
This is existing code - https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L944 .. I'm not adding anything new here..

fossedihelm replied:
I see, but this means you found a bug :)

aqilbeig replied:
:) .. yes i have updated it locally.. will push the changes post your review to avoid multiple commits.. thanks for the review and catching this

fossedihelm replied:
Would you mind creating a separate PR to fix this? So that we can backport it?

aqilbeig replied:
I was thinking to include that in this PR only... I am fine fixing that with this one if you don't mind.. i could push the commit to this PR

fossedihelm replied:
I would prefer doing it in a separate PR. At least it would be fine if you did it as the first separate commit.
This bug is present in previous stable versions, and we need to fix it.

aqilbeig replied:
@fossedihelm fixed as a separate commit https://github.com/kubevirt/kubevirt/pull/14183/commits/1a5071f88e7cf43b68d53017d860b1a59ead5a7c

aqilbeig commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,9 +840,58 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
+func (c *Controller) getUnavailableVMICount(vms []*virtv1.VirtualMachine) (int, error) {
+	unavailableCount := 0
+	for _, vm := range vms {
+		vmi, err := c.clientset.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, metav1.GetOptions{})
```
Initially, i tried that approach only; however - there is some delay between the deletion request and actual deletion happening due to which it was not fetching the right unavailable count and more VMIs getting rotated than maxUnavailable count

I have tested this multiple times with different approaches like tried with sleep of 1s; however, that also sometimes not fetching the right count and it was increasing the overall time for rollout too..

fossedihelm replied:
Well, there are several things here
- I think there is something wrong in general:
The `getUnavailableVMICount` is executed inside a for loop of `vmUpdatedList`, that is the same list that is used inside the `getUnavailableVMICount` function itself, which means $O(n^2)$
- The latter is performed every cycle and not only in edge cases
- We could at least move to `List` request instead of `Get`
- Even better, we can use the expectations

I would suggest you initially fix the first point, which is IMHO the root cause of the misbehavior.

aqilbeig replied:
> The getUnavailableVMICount is executed inside a for loop of vmUpdatedList, that is the same list that is used inside the getUnavailableVMICount function itself, which means 

yes it is $O(n^2)$ as we need to get the unavailable count before putting another update/deletion request to maintain that we don't exceed maxUnavailable count 

it is calling [opportunistic update](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/pool/pool.go#L1164C11-L1164C30) prior to proactive update which will ensure VMI update will only start once all VM update is done that means vmUpdatedList pass to proactiveUpdate will have all VMs of that particular pool. I could pass the pool object to getUnavailable function; however it would also have the same list of VMIs. Idea is to get count of unavailable VMIs of pool which is getting update

> The latter is performed every cycle and not only in edge cases

I could move this later so that this is called only if VMI is outdated

> We could at least move to List request instead of Get

the reason we need `Get` as we need to check  whether the VMI has already been in terminating or deletion state or in notReady state

aqilbeig replied:
@fossedihelm could you please check my inline response

aqilbeig replied:
we could change this to List like this - would this be fine?

```
func (c *Controller) getUnavailableVMICount1(pool *poolv1.VirtualMachinePool, vms []*virtv1.VirtualMachine) (int, error) {
    unavailableCount := 0
    
    // Create a map of VM names for quick lookup
    vmMap := make(map[string]struct{})
    for _, vm := range vms {
        vmMap[vm.Name] = struct{}{}
    }

    // Get all VMIs in the namespace
    vmiList, err := c.clientset.VirtualMachineInstance(pool.Namespace).List(context.Background(), metav1.ListOptions{})
    if err != nil {
        return 0, err
    }

    // Count VMIs that belong to VMs in this pool and are not ready
    for _, vmi := range vmiList.Items {
        if _, exists := vmMap[vmi.Name]; exists && !isVMIReady(&vmi) {
            unavailableCount++
        }
    }

    // Count VMs that don't have a VMI (they are considered unavailable)
    for _, vm := range vms {
        found := false
        for _, vmi := range vmiList.Items {
            if vmi.Name == vm.Name {
                found = true
                break
            }
        }
        if !found {
            unavailableCount++
        }
    }

    return unavailableCount, nil
}
```

fossedihelm replied:
Sorry, but I still not understand why we should cycle over all the other vmis while focusing on a single one.
Isn't enough something like this?
```golang

	unavailableCount, err := c.getUnavailableVMICount(vmUpdatedList)
	if err != nil {
		return err
	}

	if unavailableCount >= maxUnavailableInt {
		log.Log.V(4).Infof("Delaying proactive update for pool %s/%s - max unavailable (%d) reached", pool.Namespace, pool.Name, maxUnavailableInt)
		key, err := controller.KeyFunc(pool)
		if err != nil {
			return err
		}
		c.queue.AddAfter(key, defaultRetryDelay)
		return nil
	}

	for i := range vmUpdatedList {
		vm := vmUpdatedList[i]
		obj, exists, err := c.vmiStore.GetByKey(controller.NamespacedKey(vm.Namespace, vm.Name))
		if err != nil {
			return err
		}
		if !exists {
			continue
		}
		vmi := obj.(*virtv1.VirtualMachineInstance)

		updateType, err := c.isOutdatedVMI(vm, vmi)
		if err != nil {
			return err
		}
		if updateType == proactiveUpdateTypeNone {
			continue
		}

		if err := c.handleResourceUpdate(pool, vm, vmi, updateType); err != nil {
			return err
		}
	}
	return nil
	```

aqilbeig replied:
@fossedihelm let me try to explain this with an example 

We have an vmpool with 4 replicas and maxUnavailable defined as 1 that mean that during an update not more than 1 VMI can be unavailable at any point of time

Initial State:
```
Pool: web-pool (4 replicas, maxUnavailable=1)
VMs: [web-vm-1, web-vm-2, web-vm-3, web-vm-4]
```

Upgrade Process:

First Iteration:
```
// Check unavailable count
getUnavailableVMICount() returns 0 (all VMIs are ready)
// Since 0 < maxUnavailable(1), we can proceed with upgrade
// Update web-vm-1
```

After web-vm-1 update:
State:
```
- web-vm-1: VMI not ready (upgrading)
- web-vm-2,3,4: VMI ready (old version)
getUnavailableVMICount() returns 1 (web-vm-1)
// 1 == maxUnavailable(1), must wait
```

When web-vm-1 becomes ready:
State:
```
- web-vm-1: VMI ready (new version)
- web-vm-2,3,4: VMI ready (old version)
getUnavailableVMICount() returns 0
// 0 < maxUnavailable(1), can proceed
// Update web-vm-2
```


After web-vm-2 update:
State:
```
- web-vm-1: VMI ready (new version)
- web-vm-2: VMI not ready (upgrading)
- web-vm-3,4: VMI ready (old version)
getUnavailableVMICount() returns 1
// 1 == maxUnavailable(1), must wait
```

This pattern continues until all VMs are upgraded. so it will ensure that number of VMIs updating simultaneously should not be more than maxUnavailable count at any point of time during the upgrade process.

Additionally, it ensures that if an incorrect update is triggered—such as updating the VM pool with a non-existent container disk image—the system will not proceed to update all VMIs. Instead, the update will stop after the first VMI fails to become healthy or ready. This mechanism guarantees that the VM pool maintains at least three healthy VMIs, ensuring that no more than one VMI is unavailable at any given time.

aqilbeig replied:
If we go with the approach of checking the unavailableCount only before the loop, then as soon as the VM pool update is triggered, it will evaluate whether unavailableCount >= maxUnavailable. Since the update is applied to a healthy VM pool, this condition will initially be false, allowing the loop to proceed. As a result, VMIs will be updated and deletions submitted for all of them without enforcing the constraint that no more than maxUnavailable VMIs should be unavailable at any time.

aqilbeig replied:
@fossedihelm could you please check my above comments? Do let me know if it is still not clear?

fossedihelm replied:
Thanks @aqilbeig  for the explanation. Now it's clear to me what you want to achieve.
I still believe that this continuous checking is a little bit of an anti-pattern.
I have this suggestion. At the beginning of each cycle, we know how many vmi can be updated.
We can use that number as a gate for that specific controller cycle:
```golang

	unavailableCount, err := c.getUnavailableVMICount(vmUpdatedList)
	if err != nil {
		return err
	}

	if unavailableCount >= maxUnavailableInt {
		log.Log.V(4).Infof("Delaying proactive update for pool %s/%s - max unavailable (%d) reached", pool.Namespace, pool.Name, maxUnavailableInt)
		key, err := controller.KeyFunc(pool)
		if err != nil {
			return err
		}
		c.queue.AddAfter(key, defaultRetryDelay)
		return nil
	}

	maxUpdatable := maxUnavailableInt - unavailableCount
	for i := range vmUpdatedList {
		if maxUpdatable <= 0 {
			log.Log.V(4).Infof("Delaying proactive update for pool %s/%s - max unavailable (%d) reached", pool.Namespace, pool.Name, maxUnavailableInt)
			key, err := controller.KeyFunc(pool)
			if err != nil {
				return err
			}
			c.queue.AddAfter(key, defaultRetryDelay)
			return nil
		}
		vm := vmUpdatedList[i]
		obj, exists, err := c.vmiStore.GetByKey(controller.NamespacedKey(vm.Namespace, vm.Name))
		if err != nil {
			return err
		}
		if !exists {
			continue
		}
		vmi := obj.(*virtv1.VirtualMachineInstance)

		updateType, err := c.isOutdatedVMI(vm, vmi)
		if err != nil {
			return err
		}
		if updateType == proactiveUpdateTypeNone {
			continue
		}

		if err := c.handleResourceUpdate(pool, vm, vmi, updateType); err != nil {
			return err
		}
		maxUpdatable--
	}
```
WDYT?
Of course the code can be improved (code duplication).

aqilbeig commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter.go:
```diff
@@ -126,11 +129,40 @@ func ValidateVMPoolSpec(ar *admissionv1.AdmissionReview, field *k8sfield.Path, p
+				})
+			} else {
+				percentage := strings.TrimSuffix(spec.MaxUnavailable.StrVal, "%")
+				if _, err := strconv.Atoi(percentage); err != nil {
```
we are checking for <0 later below. It would be nice to have as well.. I'll update the code

fossedihelm commented:
/test pull-kubevirt-build
/test pull-kubevirt-generate
/test pull-kubevirt-code-lint
/test pull-kubevirt-verify-go-mod
/test pull-kubevirt-unit-test
/test pull-kubevirt-manifests
/test pull-kubevirt-check-unassigned-tests

fossedihelm commented:
/test pull-kubevirt-e2e-k8s-1.32-sig-compute

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter.go:
```diff
@@ -126,11 +129,40 @@ func ValidateVMPoolSpec(ar *admissionv1.AdmissionReview, field *k8sfield.Path, p
+				}
+			}
+		} else if spec.MaxUnavailable.Type == intstr.Int {
+			if spec.MaxUnavailable.IntVal < 0 {
```
Sorry for not noticing before: is 0 a valid value?  With 0 we are not tolerating anything, which will block the update, no?

aqilbeig replied:
no this won't block as i am setting this to 1 while calculating maxUnavailable:
```
	if maxUnavailableInt < 1 {
		maxUnavailableInt = 1
	}
```

aqilbeig replied:
https://github.com/kubevirt/kubevirt/pull/14183/files#diff-03f7881dab06cb6b4994a2dd1b11c77e7ebe2add843b085c1c492c5eb1ca476aR969

fossedihelm replied:
Right! Thanks. But in this sense, it is misleading: it looks like we are allowing 0 unavailability, which unfortunately is not.

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter.go:
```diff
@@ -144,10 +144,16 @@ func ValidateVMPoolSpec(ar *admissionv1.AdmissionReview, field *k8sfield.Path, p
+						Message: fmt.Sprintf("maxUnavailable percentage value %q is invalid: %v", percentage, err),
+						Field:   field.Child("maxUnavailable").String(),
+					})
+				} else if val < 0 || val > 100 {
```
Again, is 0 a valid percentage?

aqilbeig replied:
calculateMaxUnavailableInt function is taking of the same

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmpool-admitter.go:
```diff
@@ -144,10 +144,16 @@ func ValidateVMPoolSpec(ar *admissionv1.AdmissionReview, field *k8sfield.Path, p
 			} else {
 				percentage := strings.TrimSuffix(spec.MaxUnavailable.StrVal, "%")
-				if _, err := strconv.Atoi(percentage); err != nil {
+				if val, err := strconv.Atoi(percentage); err != nil {
```
the changes in this file are part of the first commit

aqilbeig replied:
@fossedihelm sorry i didn't understand this - I squash the commits to avoid multiple commits and to keep the PR commits clean.. Do i need to update the commit message or change something in this code?

fossedihelm commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -961,7 +961,7 @@ func calculateMaxUnavailableInt(pool *poolv1.VirtualMachinePool) (int, error) {
 			return 0, fmt.Errorf("invalid maxUnavailable percentage: %v", err)
 		}
-		maxUnavailableInt = int(float64(totalReplicas) * float64(percentage) / 100.0)
+		maxUnavailableInt = int(totalReplicas) * int(percentage) / 100
```
the changes in this file are part of the first commit

fossedihelm commented on pkg/virt-controller/watch/pool/pool_test.go:
```diff
@@ -116,6 +115,16 @@ var _ = Describe("Pool", func() {
 			key, err := virtcontroller.KeyFunc(vm)
 			Expect(err).To(Not(HaveOccurred()))
 			mockQueue.Add(key)
+			fakeVirtClient.Fake.PrependReactor("get", "virtualmachineinstances", func(action k8stesting.Action) (handled bool, obj runtime.Object, err error) {
```
the changes in this file are part of the first commit

fossedihelm commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -832,9 +840,58 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
+func (c *Controller) getUnavailableVMICount(vms []*virtv1.VirtualMachine) (int, error) {
+	unavailableCount := 0
+	for _, vm := range vms {
+		vmi, err := c.clientset.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, metav1.GetOptions{})
```
Maybe now we can switch to using vmistore

fossedihelm reviewed (commented):
The first and the third commit messages should be adjusted. Thanks

fossedihelm commented:
/test pull-kubevirt-build
/test pull-kubevirt-generate
/test pull-kubevirt-code-lint
/test pull-kubevirt-verify-go-mod
/test pull-kubevirt-unit-test
/test pull-kubevirt-manifests
/test pull-kubevirt-check-unassigned-tests

@aqilbeig Release note should be adjusted. Keep in mind that the release note should be
inside this tag

```
 ```release-note

```

aqilbeig commented:
> /test pull-kubevirt-build /test pull-kubevirt-generate /test pull-kubevirt-code-lint /test pull-kubevirt-verify-go-mod /test pull-kubevirt-unit-test /test pull-kubevirt-manifests /test pull-kubevirt-check-unassigned-tests
> 
> @aqilbeig Release note should be adjusted. Keep in mind that the release note should be inside this tag
> 
> ```
>  ```release-note
> ```

@fossedihelm Thank you for initiating the CI tests. The current tests have passed successfully. I’ve also updated the release notes.

fossedihelm commented on pkg/virt-controller/watch/pool/pool_test.go:
```diff
@@ -114,6 +115,16 @@ var _ = Describe("Pool", func() {
 			key, err := virtcontroller.KeyFunc(vm)
 			Expect(err).To(Not(HaveOccurred()))
 			mockQueue.Add(key)
+			fakeVirtClient.Fake.PrependReactor("get", "virtualmachineinstances", func(action k8stesting.Action) (handled bool, obj runtime.Object, err error) {
+				getAction, ok := action.(k8stesting.GetAction)
+				if !ok {
+					return false, nil, nil
+				}
+				if getAction.GetNamespace() == vm.Namespace && getAction.GetName() == vm.Name {
+					return true, vm, nil
+				}
+				return false, nil, nil
+			})
```
We don't need this :)

aqilbeig replied:
yes you are right as we are no more doing the get call ...  let me push that change.. :(

aqilbeig replied:
@fossedihelm removed..

fossedihelm reviewed (commented):
We are almost there from my POV

fossedihelm commented:
/test all

aqilbeig commented:
@fossedihelm Updated the PR.. 🙏🤞

aqilbeig commented:
@fossedihelm It seems that all the required checks has been passed successfully...

0xFelix reviewed (commented):
A few more nits, we're almost there :)

aqilbeig commented:
@0xFelix @fossedihelm Would it be okay if I create a follow-up PR? This one is starting to feel never-ending, and I think it's important to stop iterating at some point.  It's been challenging to get everything perfect in a single PR..

aqilbeig commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -887,89 +944,84 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
-				log.Log.Object(pool).Infof("Proactively updating vm %s/%s in pool via label patch", vm.Namespace, vm.Name)
-			}
-		}(i)
+	if maxUnavailableInt < 1 {
```
@0xFelix This is an optional parameter that defaults to 100% to maintain backward compatibility when it's not specified in the VMPool spec. It's mainly intended to handle cases where someone provides a percentage or value less than 1

fossedihelm replied:
I think it is ok defaulting to the minimum allowed value.

aqilbeig replied:
@0xFelix @fossedihelm would it be fine if i could resolve this conversation? Could I please get an lgtm?

aqilbeig commented:
> Apart from [#14183 (comment)](https://github.com/kubevirt/kubevirt/pull/14183#discussion_r2075329526) it can be as a follow up, sure!

@0xFelix This is an optional parameter that defaults to 100% to maintain backward compatibility when it's not specified in the VMPool spec. It's mainly intended to handle cases where someone provides a percentage or value less than 1

0xFelix commented:
Ok, I'm fine with the changes. Let's wait for @fossedihelm.

Also, @iholder101 or @vladikr, can you PTAL?

fossedihelm commented:
/lgtm

aqilbeig commented:
@mhenriks or @vladikr or @xpivarc Request you to PTAL and if looks good to you, kindly approve the PR?

jean-edouard commented:
/cc

jean-edouard reviewed (commented):
/approve
/hold
Code looks good, I'm not all that familiar with VM pools but I trust reviews from vmpool experts.
Leaving it to @0xFelix to unhold when ready.

aqilbeig commented on pkg/virt-controller/watch/pool/pool.go:
```diff
@@ -15,27 +15,28 @@ import (
 
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+
 	virtv1 "kubevirt.io/api/core/v1"
 	poolv1 "kubevirt.io/api/pool/v1alpha1"
 	"kubevirt.io/client-go/kubecli"
 	"kubevirt.io/client-go/log"
 
-	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
```
sure.. i'll create the follow up PR as soon as this is merge to address all remaining improvements

jean-edouard replied:
Why?
This is too small to really matter, but as a general rule we try not to leave broken things in one PR just because we have a plan to unbreak them in a follow up PR...

aqilbeig replied:
We already have this package imported above - it is just reformatting or am i missing something here..

jean-edouard replied:
Imports are grouped by path. In this diff, `apimachinery/patch` gets moved away from its `kubevirt.io/kubevirt/pkg` friends, which is bad practice.

aqilbeig replied:
@jean-edouard updated.. Could you please approve the PR again?

jean-edouard commented:
@aqilbeig heads-up, the unit test failure seems relevant

aqilbeig commented:
> @aqilbeig heads-up, the unit test failure seems relevant

Not sure why this is failing :( — it's the first time I've seen this issue, as it was passing initially. I'll investigate and take a closer look.

aqilbeig commented:
/test pull-kubevirt-unit-test

jean-edouard commented:
> > @aqilbeig heads-up, the unit test failure seems relevant
> 
> Not sure why this is failing :( — it's the first time I've seen this issue, as it was passing initially. I'll investigate and take a closer look.

nvm, executor pool != vm pool...
@ormergi FYI https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/14183/pull-kubevirt-unit-test/1925561011612422144

fossedihelm commented:
/lgtm
@0xFelix up to you to unhold

aqilbeig commented:
/test pull-kubevirt-e2e-k8s-1.30-sig-compute

aqilbeig commented:
Thank you @0xFelix @fossedihelm @jean-edouard  for all your help with the review .. I truly appreciate it !!
